agent/src/main/java/com/code_intelligence/jazzer/instrumentor/HookMethodVisitor.kt

```diff
@@ -69,217 +69,189 @@ private class HookMethodVisitor(
             mv.visitMethodInsn(opcode, owner, methodName, methodDescriptor, isInterface)
             return
         }
-        handleMethodInsn(HookType.BEFORE, opcode, owner, methodName, methodDescriptor, isInterface)
-    }
-
-    /**
-     * Emits the bytecode for a method call instruction for the next applicable hook type in order (BEFORE, REPLACE,
-     * AFTER). Since the instrumented code is indistinguishable from an uninstrumented call instruction, it can be
-     * safely nested. Combining REPLACE hooks with other hooks is however not supported as these hooks already subsume
-     * the functionality of BEFORE and AFTER hooks.
-     */
-    private fun visitNextHookTypeOrCall(
-        hookType: HookType,
-        appliedHook: Boolean,
-        opcode: Int,
-        owner: String,
-        methodName: String,
-        methodDescriptor: String,
-        isInterface: Boolean,
-    ) = when (hookType) {
-        HookType.BEFORE -> {
-            val nextHookType = if (appliedHook) {
-                // After a BEFORE hook has been applied, we can safely apply an AFTER hook by replacing the actual
-                // call instruction with the full bytecode injected for the AFTER hook.
-                HookType.AFTER
-            } else {
-                // If no BEFORE hook is registered, look for a REPLACE hook next.
-                HookType.REPLACE
-            }
-            handleMethodInsn(nextHookType, opcode, owner, methodName, methodDescriptor, isInterface)
-        }
-        HookType.REPLACE -> {
-            // REPLACE hooks can't (and don't need to) be mixed with other hooks. We only cycle through them if we
-            // couldn't find a matching REPLACE hook, in which case we try an AFTER hook next.
-            require(!appliedHook)
-            handleMethodInsn(HookType.AFTER, opcode, owner, methodName, methodDescriptor, isInterface)
-        }
-        // An AFTER hook is always the last in the chain. Whether a hook has been applied or not, always emit the
-        // actual call instruction.
-        HookType.AFTER -> mv.visitMethodInsn(opcode, owner, methodName, methodDescriptor, isInterface)
+        handleMethodInsn(opcode, owner, methodName, methodDescriptor, isInterface)
     }
 
     fun handleMethodInsn(
-        hookType: HookType,
         opcode: Int,
         owner: String,
         methodName: String,
         methodDescriptor: String,
         isInterface: Boolean,
     ) {
-        val hook = findMatchingHook(hookType, owner, methodName, methodDescriptor)
-        if (hook == null) {
-            visitNextHookTypeOrCall(hookType, false, opcode, owner, methodName, methodDescriptor, isInterface)
+        val matchingHooks = findMatchingHooks(owner, methodName, methodDescriptor)
```

Comment:
Could you add a few tests checking different combinations to make sure that the stack logic works?
(Ah, the `loadLibrary` hooks in `sanitizers` lead to multiple hooks on those methods)

Reply:
@fabian raised some issues on enabling multiple hooks of the same type for a given method that might steer the fuzzer in different directions. While the PR also simplifies the hook instrumentation, let's discuss how to handle this in the team